src/mcmc/gibbs.jl

```diff
@@ -11,22 +34,29 @@
 # - `GibbsContext` allows us to perform conditioning while still hit the `assume` pipeline
 #   rather than the `observe` pipeline for the conditioned variables.
 """
-    GibbsContext(target_varnames, global_varinfo, context)
+    GibbsContext{VNs}(global_varinfo, context)
 
 A context used in the implementation of the Turing.jl Gibbs sampler.
 
 There will be one `GibbsContext` for each iteration of a component sampler.
+
+`VNs` is a a tuple of symbols for `VarName`s that the current component
+sampler is sampling. For those `VarName`s, `GibbsContext` will just pass `tilde_assume`
+calls to its child context. For other variables, their values will be fixed to the values
+they have in `global_varinfo`.
+
+The naive implementation of `GibbsContext` would simply have a field `target_varnames` that
+would be a collection of `VarName`s that the current component sampler is sampling. The
+reason we instead have a `Tuple` type parameter listing `Symbol`s is to allow
+`is_target_varname` to benefit from compile time constant propagation. This is important
+for type stability of `tilde_assume`.
```

Comment:
Hmm. We sure we want to do this?

I'm happy with doing this for now, but this does seem a bit overly restrictive.

There's nothing stopping us from taking the same approach as we do in `VarInfo`, allowing both handling of the case where compile-time is possible, and also the case when it's not.

Reply:
As in, this is technically breaking functionality of `Turing.Experimental.Gibbs`, which had as one of the motivation to allow more flexible conditioning, e.g. `@varname(x[1]) => ...`, similar to `condition` and `fix`.

Reply:
Yeah, agree that this is an unfortunate regression of the abilities of the experimental Gibbs, if not of the old main Gibbs sampler. Tor and I discussed this on Slack, and the reason the regression happened was that the parts that were handling all this `target_varnames` business had to be essentially rewritten to deal with the issues with dynamic models, and thus I first went for something that works and type stable when targets are just `Symbol`s, which is the vast majority of use cases.

I agree with that this should be improved, and as, again, discussed on Slack, shouldn't be impossible to do. I would like to merge this PR though, because it does not regress non-experimental stuff, and this has already gotten too big and long-lived as a PR.

Reply:
Issue to track: https://github.com/TuringLang/Turing.jl/issues/2403